### ISSUE TITLE
std::unique_ptr<> not recommended

### ISSUE THREAD

**Issue by CJCombrink** [platisd/cpp-pimpl-tutorial#1]
I saw your video on youtube and must say it is a really good, quick and clear overview of the pImpl implementation.

I just want to comment, that from experience that `std::unique_ptr<>` is not recommended since it does not enforce the const-ness of the interface, which should be a promise/contract to the user.

See the following illustrating the issue: https://godbolt.org/z/c859zM

Currently there is no easy solution but to implement a wrapper pointer, but it should at least be mentioned.
I know there is a proposal somewhere for such a pointer but could not find the number now.
The cppreference.com website uses a `std::experimental::propagate_const<>` wrapper: https://en.cppreference.com/w/cpp/language/pimpl

**Comment by platisd** [comment]
:exploding_head: :exploding_head: 
We are looking in this with a colleague as I am typing and we are genuinely surprised it works and compiles. :sweat_smile: 

Good to know nonetheless, so thanks for the input!
To be honest I don't like the idea of having to use a `std::experimental::propagate_const<>` wrapper in the future. :confused: It looks too verbose and silly. I would expect a `unique_ptr` to preserve the const-ness of an interface, so couldn't they "fix" that instead of introducing such a wrapper? :thinking: :thinking:

**Comment by platisd** [comment]
In retrospect, I can see *why* this compiles and works, however it was not the behavior I was expecting.

**Comment by CJCombrink** [comment]
I was also caught by surprise when I discovered this.

There are alternatives:
- Qt does some macro magic: https://wiki.qt.io/D-Pointer (see Q_D) 

What I normally do is a basic wrapper: https://godbolt.org/z/f7xjr1
It is definitely not fool proof but prevents one from breaking the promise on the interface.

It would be best if this can become standard.
Here is the paper for interest sake: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4388.html

I think "fixing" the existing pointer will break a lot of code though, even though it might be broken code to begin with...

**Comment by platisd** [comment]
Cool, very interesting solutions. I will leave the issue open so more people can see this and hopefully pick up a new thing like I did! :+1:

**Comment by TamaMcGlinn** [comment]
But, simply having the [pimpl variable declared as const](https://godbolt.org/z/TKxTqd) does what you want, right?

```
std::unique_ptr<GyroscopeImpl const> mGyroscope;
```

**Comment by platisd** [comment]
> But, simply having the [pimpl variable declared as const](https://godbolt.org/z/TKxTqd) does what you want, right?
> 
> ```
> std::unique_ptr<GyroscopeImpl const> mGyroscope;
> ```

@TamaMcGlinn Yes it would, as long as you weren't planning on calling any GyroscopeImpl function that isn't const qualitified.
Ideally we'd want `const` to propagate so one could claim that any constant function of GyroscopeImpl **really** doesn't have any side effects on any any of its members' state.
